docs/LANGUAGE.md

```diff
@@ -0,0 +1,767 @@
+
+```opalog
+p[] = x :-
+  foo = 42,
+  bar = false,
+  baz = null,
+  x = {"foo": foo, "bar": [bar, baz]}
+```
+
+The result:
+
+```
+p
+# [{"foo": 42, "bar": [false, null]}]
+```
+
+By defining composite values in terms of variables and references, rules can
+define abstractions over raw data and other rules.
+
+## <a name="variables"></a> Variables
+
+Variables are another kind of term in Opalog. They can appear in both the
+head and body of rules.
+
+Variables appearing in the head of a rule can be thought of as input and
+output of the rule. For example:
+
+```opalog
+sites :- [{"name": "prod"}, {"name": "smoke1"}, {"name": "dev"}]
+q[] = name :- sites[].name = name
+```
+
+In this case, if we evaluate "q" without providing an input value for "name" we obtain the names of all of the sites:
+
+```
+q
+# ["prod", "smoke1", "dev"]
+```
+
+On the other hand, if we evaluate "q" with an input value for "name" we can determine whether "name" exists in the document defined by "q":
+
+```
+q[] = "smoke2"
+# false
+
+q[] = "dev"
+# true
+```
+
+Variables appearing in the head of a rule must also appear in a non-negated
+equality expression within the same rule. This property ensures that if the
+rule is evaluated and all of the expressions evaluate to true for some set of
+variable bindings, the variable in the head of the rule will be defined.
+
+## <a name="references"></a> References
+
+Referenced are used to access nested documents.
+
+The examples in this section use the data defined in the [Examples](#examples)
+section.
+
+The simplest reference contains no variables. For example, the following
+reference returns the hostname of the second server in the first site document
+from our example data:
+
+```
+sites[0].servers[1].hostname
+# "helium"
+```
+
+References are typically written using the "dot-access" style. The canonical form does away with "." and closely resembles dictionary lookup in a language such as Python:
+
+```
+sites[0]["servers"][1]["hostname"]
+# "helium"
+```
+
+Both forms are valid, however, the "dot-access" style is typically more readable. Note, there are two cases where brackets need to be used:
+
+1. Non-string keys such as numbers, booleans, and null.
+2. Variable keys which are described later.
+
+References are always prefixed with a variable that identifes the root
+document. In the example above this is "p". The root document may be:
+
+- a local variable.
+- a rule inside the same package.
+- a document stored in OPA.
+- a documented temporarily provided to OPA as part of a transaction.
+
+References can be used to select a value from every element in a collection.
+The following reference will select the hostnames of all the servers in our
+example data:
+
+```
+sites[].servers[].hostname
+# ["hydrogen", "helium", "lithium", "berylium", "boron"]
+```
+
+Opalog allows rules to iterate sets, arrays, and objects using the same
+square bracket syntax (e.g., "sites[]" implicitly iterates all elements in "sites").
+
```

Comment:
I think we want to make the variables that we are iterating over explicit.  And then treat the [] syntax as syntactic sugar.